chain/events/filter/index.go

```diff
@@ -236,39 +241,189 @@ func (ei *EventIndex) migrateToVersion2(ctx context.Context, chainStore *store.C
 
 	// drop the temporary indices after the migration
-	_, err = tx.Exec("DROP INDEX IF EXISTS tmp_tipset_key_cid")
+	_, err = tx.ExecContext(ctx, "DROP INDEX IF EXISTS tmp_tipset_key_cid")
 	if err != nil {
-		return xerrors.Errorf("create index tmp_tipset_key_cid: %w", err)
+		return xerrors.Errorf("drop index tmp_tipset_key_cid: %w", err)
 	}
-	_, err = tx.Exec("DROP INDEX IF EXISTS tmp_height_tipset_key_cid")
+	_, err = tx.ExecContext(ctx, "DROP INDEX IF EXISTS tmp_height_tipset_key_cid")
 	if err != nil {
 		return xerrors.Errorf("drop index tmp_height_tipset_key_cid: %w", err)
 	}
 
+	// create the final index on event.height and event.tipset_key
+	_, err = tx.ExecContext(ctx, createIndexEventHeightTipsetKey)
+	if err != nil {
+		return xerrors.Errorf("create index height_tipset_key: %w", err)
+	}
+
+	// increment the schema version to 2 in _meta table.
+	_, err = tx.ExecContext(ctx, "INSERT OR IGNORE INTO _meta (version) VALUES (2)")
+	if err != nil {
+		return xerrors.Errorf("increment _meta version: %w", err)
+	}
+
 	err = tx.Commit()
 	if err != nil {
 		return xerrors.Errorf("commit transaction: %w", err)
 	}
 
-	// during the migration, we have likely increased the WAL size a lot, so lets do some
+	ei.vacuumDBAndCheckpointWAL(ctx)
+
+	log.Infof("successfully migrated event index from version 1 to version 2 in %s", time.Since(now))
+
+	return nil
+}
+
+func (ei *EventIndex) vacuumDBAndCheckpointWAL(ctx context.Context) {
+	// During the large migrations, we have likely increased the WAL size a lot, so lets do some
 	// simple DB administration to free up space (VACUUM followed by truncating the WAL file)
-	// as this would be a good time to do it when no other writes are happening
+	// as this would be a good time to do it when no other writes are happening.
 	log.Infof("Performing DB vacuum and wal checkpointing to free up space after the migration")
-	_, err = ei.db.Exec("VACUUM")
+	_, err := ei.db.ExecContext(ctx, "VACUUM")
 	if err != nil {
 		log.Warnf("error vacuuming database: %s", err)
 	}
-	_, err = ei.db.Exec("PRAGMA wal_checkpoint(TRUNCATE)")
+	_, err = ei.db.ExecContext(ctx, "PRAGMA wal_checkpoint(TRUNCATE)")
 	if err != nil {
 		log.Warnf("error checkpointing wal: %s", err)
 	}
+}
+
+// migrateToVersion3 migrates the schema from version 2 to version 3 by creating two indices:
+// 1) an index on the event.emitter_addr column, and 2) an index on the event_entry.key column.
+func (ei *EventIndex) migrateToVersion3(ctx context.Context) error {
+	now := time.Now()
+
+	tx, err := ei.db.BeginTx(ctx, nil)
+	if err != nil {
+		return xerrors.Errorf("begin transaction: %w", err)
+	}
+	defer func() { _ = tx.Rollback() }()
 
-	log.Infof("Successfully migrated events to version 2 in %s", time.Since(now))
+	// create index on event.emitter_addr.
+	_, err = tx.ExecContext(ctx, "CREATE INDEX IF NOT EXISTS event_emitter_addr ON event (emitter_addr)")
+	if err != nil {
+		return xerrors.Errorf("create index event_emitter_addr: %w", err)
```

Comment:
Why are we changing code in prior migrations ? Is the existing idea of recreating everything so the indices also get recreated bad ?

Reply:
Becase:
* this index is created for a column that will no longer exist once this PR lands, i.e. `emitter_addr`
* the `ddls` slice always points to the latest schema, i.e. the one that replaces `emitter_addr` column with `emitter` actor ID.

Therefore, we change the index creation statement that made sense for v3 here, where the `ddls` slice has the new version, i.e. index over `emitter` column.

Reply:
> Can we just drop this index from this migration?

Not sure I follow. That index is the whole point of [Version 3 migration](https://github.com/filecoin-project/lotus/commit/95fb198597b8d996b396281da8c975f5c9c858b6).

Version 4, i.e. the changes in this PR, [drop it](https://github.com/filecoin-project/lotus/blob/ba2e7a19b8d911ecb945a41e87d0c9df39f908e6/chain/events/filter/index.go#L400) first since then [recreate it](https://github.com/filecoin-project/lotus/blob/ba2e7a19b8d911ecb945a41e87d0c9df39f908e6/chain/events/filter/index.go#L410) on the newly populated `emitter` column if that's what you are suggesting?

Reply:
I suppose that once this code goes live, then the only pathway leads up to a v4 schema anyway and you have a "drop if exists" for this index in there, so removing it entirely from a v3 migration when v4 is live is basically a noop as far as the end result is concerned. If you run this code on a v2 (which I think 1.26.0 is), you'll end up with a v4 anyway and this index will have been temporarily added then removed to get there.

Reply:
I am not sure if it is worth making optimisations here. Instead, my preference is to execute every migration unless the schema does not exist. That simple rule would make it a lot easier to reason about stateful migrations in general.

Unless there is significant performance overhead, my sense is to stick to the above migration rule.

Reply:
I'm more concerned about keeping the code clean and easy to read. I'm not a fan of moving this SQL statement into this migration as a string literal and then just tossing the result.

Reply:
IMHO reasoning about immutable migrations is much easier than trying to optimise it per new migration. 
But it looks like I am out voted here. So I i'll drop the index migration at v3 in favour of combining it into v4.